server/chatroom.py

```diff
@@ -20,36 +23,188 @@ def get_msg_update(cur, chatroom_id, date_time):
+        return result
+    except BaseException as e:
+        print(f'Exception: {e}')
+        raise HTTPException(
+            status_code=500,
+            detail="Database Error",
+        )
 
 
 # Save message functionality
-def saveMessage(cur, sender_id, chatroomID, message_sent):
+def save_message(cur, sender_id, chatroomID, message_sent):
     try:
         date_time_sent = datetime.datetime.now()
         cur.execute(
             "INSERT INTO message(sender_id, chatroom_id, message_text, date_time_sent) VALUES(%(sender_id)s, %(chatroomID)s, %(message_sent)s, %(date_time_sent)s)",
             {"sender_id": sender_id, "chatroomID": chatroomID, "message_sent": message_sent,
              "date_time_sent": date_time_sent})
         return True
-    except:
-        return False
+    except BaseException as e:
+        print(f'Exception: {e}')
+        raise HTTPException(
+            status_code=500,
+            detail="Database Error",
+        )
 
 
 # Create non-course chatroom
-def createChatroom(cur, user_id, chatroom_name, uni_id):
+def create_chatroom(cur, user_id, chatroom_name, uni_id):
     try:
-        cur.execute("INSERT INTO chatroom(chatroom_name, uni_id) VALUES(%(chatroom_name)s, %(uni_id)s)",
-                    {"chatroom_name": chatroom_name, "uni_id": uni_id})
-        cur.execute("SELECT id FROM chatroom where chatroom_name = %(chatroom_name)s   AND uni_id = %(uni_id)s",
-                    {"chatroom_name": chatroom_name, "uni_id": uni_id})
+        # Generate invite id
+        invite_id = ''.join(random.choices(string.ascii_uppercase + string.digits, k=10))
+        
+        cur.execute("INSERT INTO chatroom(chatroom_name, uni_id, invite_id) VALUES(%(chatroom_name)s, %(uni_id)s, %(invite_id)s)", {"chatroom_name" : chatroom_name, "uni_id" : uni_id, "invite_id" : invite_id})
+        
+        cur.execute("SELECT id FROM chatroom where chatroom_name = %(chatroom_name)s   AND uni_id = %(uni_id)s", {"chatroom_name" : chatroom_name, "uni_id" : uni_id})
+        
         chatroom_id = cur.fetchall()[0][0]
+        
         cur.execute(
             "INSERT INTO in_chatroom(student_id, uni_id, chatroom_id) VALUES(%(user_id)s, %(uni_id)s, %(chatroom_id)s)",
             {"user_id": user_id, "uni_id": uni_id, "chatroom_id": chatroom_id})
+        
+        return True
+    except BaseException as e:
+        print(f'Exception: {e}')
+        raise HTTPException(
+            status_code=500,
+            detail="Database Error",
+        )
+    
+#Generate invite link for student (for non-course chatroom)
+def generate_invite(cur, target_user_id, invite_sender_id, chatroom_id, uni_id):
+    try:
+
+        #Generate invite id
+        invite_id = ''.join(random.choices(string.ascii_uppercase + string.digits, k=10))
+        
+        #make sure that the university actually exists in the database
+        cur.execute("SELECT id FROM university where id = %(uni_id)s", {"uni_id" : uni_id})
```

Comment:
use a series of join to replace the multiple checks here. sth like this (i havent check the schema while writing, u prob need to debug this):
```
select s.id, t.id
from (student as s, student as t) join in_chatroom as ic using (s.student_id) join univeristy as u on (s.uni_id=u.uni_id and t.uni_id=u.uni_id) join chatroom as c using (chatroom_id)
where s.student_id=%(sender_id)s and t.student_id=%(target_id) and ic.chatroom_id=%(chatroom_id)s and c.course_id is null
```